sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/pointer/CommonPointerLibraries.java

```diff
@@ -116,6 +142,84 @@ static boolean isMemberModifiable(LLVMPointerImpl receiver, String ident,
+    }
+
+    @ExportMessage
+    static class InvokeMember {
+        @SuppressWarnings("unused")
+        @Specialization(guards = {"asClazz(receiver)==clazz", "member.equals(methodName)", "argCount==arguments.length"})
+        static Object doCached(LLVMPointerImpl receiver, String member, Object[] arguments,
+                        @CachedContext(LLVMLanguage.class) LLVMContext context, @CachedLibrary(limit = "5") InteropLibrary interop,
+                        @Cached(value = "asClazz(receiver)", allowUncached = true) LLVMInteropType.Clazz clazz,
+                        @Cached(value = "clazz.findMethodByArguments(receiver, member, arguments)", allowUncached = true) Method method,
+                        @Cached(value = "arguments.length", allowUncached = true) int argCount,
+                        @Cached(value = "method.getName()", allowUncached = true) String methodName,
+                        @Cached(value = "getLLVMFunction(context, method, clazz, member)", allowUncached = true) LLVMFunction llvmFunction,
+                        @Cached(value = "create(llvmFunction)", allowUncached = true) LLVMAccessSymbolNode accessSymbolNode)
+                        throws UnsupportedMessageException, ArityException, UnsupportedTypeException {
+            Object[] newArguments = addSelfObject(receiver, arguments);
+            return interop.execute(accessSymbolNode.execute(), newArguments);
+        }
+
+        @SuppressWarnings("unused")
+        @Specialization
+        static Object doResolve(LLVMPointerImpl receiver, String member, Object[] arguments,
+                        @CachedContext(LLVMLanguage.class) LLVMContext context, @CachedLibrary(limit = "5") InteropLibrary interop,
+                        @Cached(value = "asClazz(receiver)", allowUncached = true) LLVMInteropType.Clazz clazz,
+                        @Cached(value = "clazz.findMethodByArguments(receiver, member, arguments)", allowUncached = true) Method method,
+                        @Cached(value = "arguments.length", allowUncached = true) int argCount,
+                        @Cached(value = "method.getName()", allowUncached = true) String methodName,
+                        @Cached(value = "getLLVMFunction(context, method, clazz, member)", allowUncached = true) LLVMFunction llvmFunction,
+                        @Cached(value = "create(llvmFunction)", allowUncached = true) LLVMAccessSymbolNode accessSymbolNode)
+                        throws UnsupportedMessageException, ArityException, UnsupportedTypeException, UnknownIdentifierException {
+            Object[] newArguments = addSelfObject(receiver, arguments);
+            LLVMInteropType.Clazz newClazz = asClazz(receiver);
+            Method newMethod = newClazz.findMethodByArguments(receiver, member, arguments);
+            LLVMFunction newLLVMFunction = getLLVMFunction(context, newMethod, newClazz, member);
+            Object newReceiver = context.createFunctionDescriptor(newLLVMFunction);
```

Comment:
We shouldn't create a new function descriptor here. I wonder whether we need a variant of `LLVMAccessSymbolNode` that can take a dynamic symbol as argument... /cc @Palez

Reply:
Currently the `LLVMAccessSymbolNode` must take a symbol in order to be able to access it. We could just simple create a `LLVMDynAccessSymbolNode` for this use case, something like:

`public abstract class LLVMDynAccessSymbolNode extends LLVMExpressionNode {
   public LLVMDynAccessSymbolNode() {
    }
    public abstract LLVMPointer execute();
    @Specialization
    LLVMPointer doAccess(LLVMSymbol symbol,
                    @CachedContext(LLVMLanguage.class) LLVMContext context) {
        CompilerAsserts.partialEvaluationConstant(symbol);
        if (symbol.hasValidIndexAndID()) {
            int bitcodeID = symbol.getBitcodeID(false);
            if (context.symbolTableExists(bitcodeID)) {
                AssumedValue<LLVMPointer>[] symbols = context.findSymbolTable(bitcodeID);
                int index = symbol.getSymbolIndex(false);
                AssumedValue<LLVMPointer> assumedValue = symbols[index];
                if (assumedValue != null) {
                    return assumedValue.get();
                }
            }
        }
        CompilerDirectives.transferToInterpreter();
        throw new LLVMLinkerException(this, String.format("External %s %s cannot be found.", symbol.getKind(), symbol.getName()));
    }
}`

@pichristoph Would you like me to add this class onto the repo or would you like to do it?

Reply:
@Palez if it is fine for you, I will add it. Thank you for the suggestion/solution!